usr/lib/cca_stdll/cca_specific.c

```diff
@@ -3815,3 +3815,374 @@ CK_RV token_specific_generic_secret_key_gen(STDLL_TokData_t * tokdata,
+        *wrapped_key_len = buffer_len;
+        return CKR_OK;
+    }
+
+    if ((CK_ULONG)buffer_len > *wrapped_key_len) {
+        *wrapped_key_len = buffer_len;
+        return CKR_BUFFER_TOO_SMALL;
+    }
+
+    memcpy(wrapped_key, buffer, buffer_len);
+    *wrapped_key_len = buffer_len;
+
+    return CKR_OK;
+}
+
+static CK_RV ccatok_unwrap_key_rsa_pkcs(CK_MECHANISM *mech,
+                                        OBJECT *wrapping_key, OBJECT *key,
+                                        CK_BYTE *wrapped_key,
+                                        CK_ULONG wrapped_key_len)
+{
+    long return_code, reason_code, rule_array_count;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0 };
+    CK_BYTE buffer[3500] = { 0, };
+    CK_BYTE dummy[AES_KEY_SIZE_256] = { 0, };
+    long buffer_len = sizeof(buffer);
+    CK_ATTRIBUTE *attr, *wrap_key_opaque,*key_opaque = NULL;
+    CK_ATTRIBUTE *value = NULL, *value_len = NULL;
+    CK_OBJECT_CLASS key_class;
+    CK_KEY_TYPE key_type, cca_key_type;
+    CK_ULONG key_size = 0;
+    uint16_t val;
+    CK_RV rc;
+
+    if (!template_attribute_find(key->template, CKA_CLASS, &attr))
+         return CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT;
+    key_class = *(CK_OBJECT_CLASS *)attr->pValue;
+
+    if (key_class != CKO_SECRET_KEY)
+        return CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT;
+
+    if (!template_attribute_find(key->template, CKA_KEY_TYPE, &attr))
+        return CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT;
+    key_type = *(CK_KEY_TYPE *) attr->pValue;
+
+    switch (key_type) {
+    case CKK_DES:
+    case CKK_DES2:
+    case CKK_DES3:
+        rule_array_count = 2;
+        switch (mech->mechanism) {
+        case CKM_RSA_PKCS:
+            memcpy(rule_array, "DES     PKCS-1.2", 2 * CCA_KEYWORD_SIZE);
+            break;
+        default:
+            return CKR_MECHANISM_INVALID;
+        }
+        break;
+    case CKK_AES:
+        rule_array_count = 2;
+        switch (mech->mechanism) {
+        case CKM_RSA_PKCS:
+            memcpy(rule_array, "AES     PKCS-1.2", 2 * CCA_KEYWORD_SIZE);
+            break;
+        default:
+            return CKR_MECHANISM_INVALID;
+        }
+        break;
+    default:
+        return CKR_KEY_NOT_WRAPPABLE;
+    }
+
+    if (!template_attribute_find(wrapping_key->template, CKA_IBM_OPAQUE,
+                                 &wrap_key_opaque))
+        return CKR_KEY_NOT_WRAPPABLE;
+
+    dll_CSNDSYI(&return_code, &reason_code, NULL, NULL, &rule_array_count,
+                rule_array, (long *)&wrapped_key_len, wrapped_key,
+                (long *)&wrap_key_opaque->ulValueLen, wrap_key_opaque->pValue,
+                &buffer_len, buffer);
+
+    if (return_code != CCA_SUCCESS) {
+        TRACE_ERROR("CSNDSYI (SYMMETRIC KEY IMPORT) failed."
+                    " return:%ld, reason:%ld\n", return_code, reason_code);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    if (buffer[0] != 0x01) { /* Internal key token */
+        TRACE_DEVEL("key token invalid\n");
```

Comment:
ERROR trace please

Reply:
Don't confuse DEVEL trace with DEBUG trace. DEVEL traces are enabled with the trrace level of 4, which is available by default. Only for DEBUG traces it needs to be recompiled.